Dataclass inheritance doesn't work when base class is decoded first

### Description

When using inheritance in dataclasses, msgspec decodes a derived class as it's base class when the base class is decoded once beforehand. I've made an minimal example to replicate the issue.

`a.py`:

```python
from dataclasses import dataclass


@dataclass
class Base:
    a: int


@dataclass
class Foo(Base):
    b: int
```

`b.py`:
```python
import msgspec


def test_decode():
    # Base class decoded first
    from a import Base

    x = Base(a=1)
    ser = msgspec.json.encode(x)
    dser = msgspec.json.decode(ser, type=Base)
    assert x == dser

    from a import Foo

    x = Foo(a=1, b=2)
    ser = msgspec.json.encode(x)
    dser = msgspec.json.decode(ser, type=Foo)
    assert x == dser


if __name__ == "__main__":
    test_decode()
```

Running `pytest`:
```shell
pytest b.py
```
```
===================================================== test session starts =====================================================
platform linux -- Python 3.11.6, pytest-7.4.3, pluggy-1.3.0
rootdir: /tmp/
plugins: anyio-4.0.0
collected 2 items

b.py F                                                                                                                  [100%]
========================================================== FAILURES ===========================================================
_________________________________________________________ test_decode _________________________________________________________

    def test_decode():
        from a import Base

        x = Base(a=1)
        ser = msgspec.yaml.encode(x)
        dser = msgspec.yaml.decode(ser, type=Base)
        assert x == dser

        from a import Foo

        x = Foo(a=1, b=2)
        ser = msgspec.yaml.encode(x)
        dser = msgspec.yaml.decode(ser, type=Foo)
>       assert x == dser
E       assert Foo(a=1, b=2) == Base(a=1)

b.py:20: AssertionError
=================================================== short test summary info ===================================================
FAILED b.py::test_decode - assert Foo(a=1, b=2) == Base(a=1)
====================================================== 1 failed in 0.04s ======================================================
```

However, if the derived class is decoded first then the base class is correctly decoded. This works perfectly:

```python
def test_decode():
    # derived class decoded first
    from a import Foo

    x = Foo(a=1, b=2)
    ser = msgspec.yaml.encode(x)
    dser = msgspec.yaml.decode(ser, type=Foo)
    assert x == dser

    from a import Base

    x = Base(a=1)
    ser = msgspec.yaml.encode(x)
    dser = msgspec.yaml.decode(ser, type=Base)
    assert x == dser
```

Tested with:
 - Python 3.11.6 and 3.9.18
 - msgspec 0.18.4

--- a/msgspec/_core.c
+++ b/msgspec/_core.c
@@ -3706,7 +3706,7 @@ error:
 
 static bool
 get_msgspec_cache(MsgspecState *mod, PyObject *obj, PyTypeObject *type, PyObject **out) {
-    PyObject *cached = PyObject_GetAttr(obj, mod->str___msgspec_cache__);
+    PyObject *cached = PyObject_GenericGetAttr(obj, mod->str___msgspec_cache__);
     if (cached != NULL) {
         if (Py_TYPE(cached) != type) {
             Py_DECREF(cached);

--- a/tests/test_common.py
+++ b/tests/test_common.py
@@ -2582,6 +2582,20 @@ class TestDataclass:
         assert dec.decode(proto.encode(msg)) == msg
         assert dec2.decode(proto.encode(msg)) == msg
 
+    def test_decode_dataclass_subclasses(self, proto):
+        @dataclass
+        class Base:
+            x: int
+
+        @dataclass
+        class Sub(Base):
+            y: int
+
+        msg = proto.encode({"x": 1, "y": 2})
+
+        assert proto.decode(msg, type=Base) == Base(1)
+        assert proto.decode(msg, type=Sub) == Sub(1, 2)
+
     def test_multiple_dataclasses_errors(self, proto):
         @dataclass
         class Ex1:

